### ISSUE TITLE
Freeze in page_already_present

### ISSUE THREAD

**Issue by mrks** [LLNL/umap#109]
I am trying to run the `psort` example, however, the program freezes either when calling `uunmap` or -when `UMAP_BUFSIZE=10`- during initialization. I tracked it down to 

```
2699597:2699599 [DEBUG][../src/umap/Buffer.cpp:255]: page_already_present Waiting for state: (ANY), { 0x7f1418370000, LEAVING, DIRTY }
```

The `pthread_cond_wait` is never completed.

My kernel is: `Linux nemea 5.8.0-28-generic #30-Ubuntu SMP Thu Nov 5 13:24:33 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

Any idea where to look? I tried entirely new systems, but was still unable to run your example. Could this be an OS issue? What configuration is known to be good?

**Comment by mrks** [comment]
Here is a shorter example, which does not take the code path described above, but also freezes:

```cpp
#include <cassert>
#include <cstring>
#include <iostream>
#include <fstream>
#include <fcntl.h>
#include <string>
#include <umap/umap.h>

int main() {
  const auto size = 4096;

  std::ofstream ofs("test", std::ios::binary | std::ios::out);
  std::cout << __LINE__ << std::endl;
  ofs.seekp((size) - 1);
  std::cout << __LINE__ << std::endl;
  ofs.write("", 1);
  std::cout << __LINE__ << std::endl;
  ofs.close();
  std::cout << __LINE__ << std::endl;

  auto fd = open("test", O_RDWR | O_LARGEFILE | O_DIRECT | O_CREAT, S_IRUSR | S_IWUSR);
  std::cout << __LINE__ << std::endl;
  auto base_addr = umap(NULL, size, PROT_READ|PROT_WRITE, UMAP_PRIVATE, fd, 0);
  std::cout << __LINE__ << std::endl;
  strcpy(reinterpret_cast<char*>(base_addr), "Hi, this is a test");
  std::cout << __LINE__ << std::endl;
  uunmap(base_addr, size);
  std::cout << __LINE__ << std::endl;
  close(fd);
  std::cout << __LINE__ << std::endl;
}
```

The execution never makes it past the `uunmap`:

```
13
15
17
19
22
2708181:2708181 [DEBUG][../src/umap/umap.cpp:30]: umap region_addr: 0, region_size: 4096, prot: 3, flags: 2, offset: 0
2708181:2708181 [DEBUG][../src/umap/RegionManager.cpp:232]: set_umap_page_size Adjusting page size from 0 to 4096
2708181:2708181 [DEBUG][../src/umap/RegionManager.cpp:208]: set_max_pages_in_buffer Maximum pages in page buffer changed from 0 to 339959117 pages
2708181:2708181 [INFO][../src/umap/umap.cpp:138]: umap_ex region_addr: 0, region_size: 4096, prot: 3, flags: 2, offset: 0, store: 0, umap_psize: 4096
2708181:2708181 [DEBUG][../src/umap/store/StoreFile.cpp:21]: StoreFile region: 0x7f6f01bac000 rsize: 4096 alignsize: 4096 fd: 3
2708181:2708181 [DEBUG][../src/umap/RegionManager.cpp:43]: addRegion No active regions, initializing engine
2708181:2708181 [DEBUG][../src/umap/Uffd.cpp:147]: Uffd
 maximum fault events: 256
            page size: 4096
2708181:2708181 [DEBUG][../src/umap/WorkerPool.hpp:73]: start_thread_pool Starting Uffd Manager Pool of 1 threads
2708181:2708181 [DEBUG][../src/umap/WorkerPool.hpp:73]: start_thread_pool Starting Fill Workers Pool of 224 threads
2708181:2708181 [DEBUG][../src/umap/WorkerPool.hpp:73]: start_thread_pool Starting Evict Workers Pool of 224 threads
2708181:2708181 [DEBUG][../src/umap/WorkerPool.hpp:73]: start_thread_pool Starting Evict Manager Pool of 1 threads
2708181:2708181 [DEBUG][../src/umap/RegionManager.cpp:53]: addRegion region: 0x7f6f01bac000 - 0x7f6f01bad000, region_size: 4096, number of regions: 2
2708181:2708181 [DEBUG][../src/umap/Uffd.cpp:259]: register_region Registering 1 pages from: 0x7f6f01bac000 - 0x7f6f01bacfff
24
2708181:2708182 [DEBUG][../src/umap/Buffer.cpp:214]: process_page_event NEW: { 0x7f6f01bac000, FILLING, DIRTY } From: { m_size: 339959117, m_waits_for_avail_pd: 0, m_present_pages.size():  1, m_free_pages.size(): 339959116, m_busy_pages.size():  1 }
2708181:2708184 [DEBUG][../src/umap/FillWorkers.cpp:42]: FillWorker : { page_desc: { 0x7f6f01bac000, FILLING, DIRTY }, type: NONE } { m_size: 339959117, m_waits_for_avail_pd: 0, m_present_pages.size():  1, m_free_pages.size(): 339959116, m_busy_pages.size():  1 }
2708181:2708184 [DEBUG][../src/umap/store/StoreFile.cpp:30]: read_from_store pread(fd=3, buf=0x7f6c70001000, nb=4096, off=0)
26
2708181:2708181 [DEBUG][../src/umap/umap.cpp:43]: uunmap addr: 0x7f6f01bac000, length: 4096
2708181:2708181 [DEBUG][../src/umap/RegionManager.cpp:72]: removeRegion region: 0x7f6f01bac000 - 0x7f6f01bad000, region_size: 4096, number of regions: 1
2708181:2708181 [DEBUG][../src/umap/EvictManager.cpp:49]: EvictAll Entered
2708181:2708181 [DEBUG][../src/umap/Buffer.cpp:110]: evict_oldest_page Normal Page: { 0x7f6f01bac000, PRESENT, DIRTY }
2708181:2708181 [DEBUG][../src/umap/Buffer.cpp:333]: wait_for_page_state Waiting for state: PRESENT, { 0x7f6f01bac000, PRESENT, DIRTY }
2708181:2708181 [DEBUG][../src/umap/EvictManager.cpp:52]: EvictAll evicting: { 0x7f6f01bac000, LEAVING, DIRTY }
2708181:2708407 [DEBUG][../src/umap/EvictWorkers.cpp:26]: EvictWorker  { page_desc: { 0x7f6f01bac000, LEAVING, DIRTY }, type: FAST_EVICT } { m_size: 339959117, m_waits_for_avail_pd: 0, m_present_pages.size():  1, m_free_pages.size(): 339959116, m_busy_pages.size():  0 }
2708181:2708407 [DEBUG][../src/umap/store/StoreFile.cpp:48]: write_to_store pwrite(fd=3, buf=0x7f6f01bac000, nb=4096, off=0)
```

This does not happen if the `strcpy` is commented out. As such, it appears to be related to the handling of dirty pages.

**Comment by mayagokhale** [comment]
Thanks for reporting. We will take a look. Have you tried the 5.7 kernel or later than 5.9?

**Comment by mrks** [comment]
Thank you for your quick response and for making me look into the kernel versions. I took a couple of Equinix instances and played around:

| Ubuntu    | Kernel                                                      | Result                                                          |
|-----------|-------------------------------------------------------------|-----------------------------------------------------------------|
| 20.04 LTS | original 5.4.0-26-generic #30-Ubuntu (5.4.34)               | umap_ex only PROT_READ is supported in UMAP_RO_MODE compilation |
| 20.04 LTS | manually upgraded to 5.7.0-050700-generic #202006082127     | works *                                                         |
| 20.04 LTS | manually upgraded to 5.8.0-050800-generic #202008022230     | freezes *                                                       |
| 20.04 LTS | manually upgraded to 5.9.0-050900-generic #202010112230     | works *                                                         |
| 20.10     | original 5.8.0-26-generic #27-Ubuntu (5.8.14)               | freezes                                                         |
| 20.10     | apt upgrade to 5.8.0-29-generic #31-Ubuntu (5.8.14)         | freezes                                                         |
| 20.10     | apt upgrade (PPA) to 5.8.0-32-generic #34-Ubuntu (5.8.18)   | freezes                                                         |
| 20.10     | manually upgraded to 5.9.0-050900-generic #202010112230     | works                                                           |
| 20.10     | manually upgraded to 5.10.0-051000rc6-generic #202011291930 | works                                                           |

For runs marked with an asterisk, I included the `userfaultfd.h` header of the new kernel with its absolute path. GCC had linking issues similar to what is described [here](https://github.com/seemoo-lab/owl/issues/32), so I fixed the `__packed` issue in `/usr/src`. Don't look at me like that... ;)

**Comment by IBPeng** [comment]
Another good starting point for trying UMap is https://github.com/LLNL/umap-apps/blob/develop/regression-tests/regression_tests_run.sh . It includes sort and other test cases, and we know the configurations and expected output.

**Comment by egreen77** [comment]
Just to add a bit more detail, this is unfortunately a known bug resulting from other changes introduced in in kernel 5.8. We were tracking a patch series related this issue (https://patchwork.kernel.org/project/linux-mm/patch/20200811183950.10603-1-peterx@redhat.com/) and indeed it does seem to have been fixed in the 5.9 release.

**Comment by mrks** [comment]
Thank you for providing the context. May I suggest including a note about the limited support for 5.8 somewhere in the documentation?
